designs/sam-app-config.md

```diff
@@ -0,0 +1,201 @@
+SAM CLI App Level Config
+====================================
+
+
+What is the problem?
+--------------------
+
+Today users of SAM CLI need to invoke the CLI directly with all parameters supplied to its commands.
+
+for e.g: `sam build --use-container --debug`
+
+But often, during the lifecycle of building and deploying a serverless application. the same commands get run repeatedly to build, package and deploy, before solidifying into the final application. 
+
+These CLI commands are often long and have many changing parts.
+
+Have a look at the following series of workflows
+
+
+* `sam build --use-container --template ... --parameter-overrides=... --skip-pull-image --manifest ...`
+
+* `sam package --s3-bucket ... --template-file ... --output-template-file ... --s3-prefix ... --kms-key-id ...`
+
+* `sam deploy deploy --template-file ... --stack-name ... --capabilities ... --tags ... --parameter-overrides ... --kms-key-id ...`
+
+If this could be condensed into a series of workflows that look like
+
+* `sam build`
+* `sam package`
+* `sam deploy`
+
+That would be a huge user experience win.
+
+What will be changed?
+---------------------
+
+The suite of commands supported by SAM CLI would be aided by looking for a configuration file thats locally located to the template.yaml by default. 
+
+`.aws-sam/sam-app-config`
+
+This configuration would solely be used for specifiying the parameters that each of SAM CLI commands use and would be in TOML format.
+
+A configuration file can also optionally be specified by each of the commands themselves as well.
+
+`sam build --config .aws-sam/sam-app-dev-config`
+
+Sample configuration file
+
+```
```

Comment:
INI format?

Reply:
Also, do we want to consider user profiles in a way that aligns with AWS CLI?

Reply:
this is toml. looks similair to ini but understands types better. pip today respects `pyproject.toml`. I'd rather have multiple configuration files, similar to how parameter overrides usually denote different environments.

Reply:
We should make sure toml supports what we want. I am ok with it but I prefer us to not support multiple configuration files. The upkeep on that might be more than we originally think.